hal-qspi test

maciejbocianski opened the pull request:
### Description

This PR adds:
- QSPI hal test
- temporary fix for `qspi_free` return value for STM QSPI implementation
- qspi pin names for DISCO_L475VG_IOT01A target
- fix for QSPI custom command sending for NRF target
- fix for qspi R/W opcodes mapping for NRF target

Supported targets: `DISCO_L475VG_IOT01A`, `NRF52840_DK`

Known issues:
- STM: lack of `qspi_free` implementation
- STM: bug with to early write finish indication on DISCO_L475VG_IOT01A target (for more details see `STM_DISCO_L475VG_IOT01A_WRITE_4IO_BUG_WORKAROUND`)
- STM: `HAL_QSPI_Init` is failing on ARM compiler with `HAL_BUSY` error
- STM: malloc is failing on DISCO_L475VG_IOT01A target on GCC (every dynamic memory allocation returns NULL)



### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

maciejbocianski committed:
STM qspi: temporary fix for qspi_free return value

maciejbocianski committed:
STM: add qspi pin names for DISCO_L475VG_IOT01A

maciejbocianski committed:
NRF5: fix qspi custom command sending

maciejbocianski committed:
NRF5: fix qspi R/W opcodes mapping

maciejbocianski committed:
add hal-qspi test

jamesbeyond commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -192,7 +192,8 @@ qspi_status_t qspi_init(qspi_t *obj, PinName io0, PinName io1, PinName io2, PinN
     // TODO
-    return QSPI_STATUS_ERROR;
+    //return QSPI_STATUS_ERROR;
+    return QSPI_STATUS_OK;
```
What's the reason, this function can't be fully implemented but required a temperate fix?

0xc0170 replied:
I think it's temporary fix. It should be implemented

@adustm Please review

offirko commented on TESTS/mbed_hal/qspi/qspi_test_utils.cpp:
```diff
@@ -0,0 +1,283 @@
+{
+    uint8_t reg_data[QSPI_STATUS_REGISTER_SIZE];
+
+    reg_data[0] = STATUS_BIT_QE;
```
SFDP standard defines Quad Enable procedure depending on QER  bits value.
This procedure of matches QER setup 010b where QE is bit 6 of status register.
Do we require a more general approach that handles Quad Enable for all QER value scenarios?

maciejbocianski replied:
Good point, could be done in next step.
But not sure if we really need such flexibility as SFDP provides in test layer.
The idea of QSPI tests was to check basic features, common for all targets/memory chips,
just to be sure that HAL API implementations works
@0xc0170 @SenRamakri

0xc0170 replied:
@offirko This is utils functions for the driver test that has no SFDP knowledge. SFDP will be part of the device block driver (tested there). What would you suggest here?

offirko replied:
AFAIU (regardless of SFDP), QE bit is frequently bit number 6 of Status Register - but not for all devices. 'quad_enable' will pass successfully because it verifies that bit 6 value has changed, but if further tests expect that quad_enable will enable QPI or FAST Quad mode they might fail for certain devices.  We can perhaps start with this hard coded value and later on move to it to per-target configuration if required.

SenRamakri commented on TESTS/mbed_hal/qspi/qspi_test_utils.cpp:
```diff
@@ -0,0 +1,283 @@
+    return ((reg[0] & STATUS_BIT_WEL) == 0 ? QSPI_STATUS_OK : QSPI_STATUS_ERROR);
+}
+
+WEAK void log_security_register(Qspi &qspi)
```
What's the purpose of these log_xx_register functions? Why are they weak? Usually we don't have to print lot of data from tests, may be you need to confine them to debug builds only.

maciejbocianski replied:
Yes, it was for debugging purpose, I'll remove it.
All weak function can be overridden if specific target require different handling

maciejbocianski commented:
PR has been rebased and updated:
- log_xx_register functions removed

cmonr commented:
If all of the reviewers could re-review:
@0xc0170 @SenRamakri @jamesbeyond @offirko @kegilbert

maciejbocianski committed:
hal-qspi test refactoring

Main idea of introduced changes is to ease adding support for new flash chips
Major changes:
- move implementation of all memory chip specific functions to memory config file (no weak functions)
- add support for 1-2-2 write

maciejbocianski commented:
Test code was updated. Changes was introduced as background for new memory chips support.
Please review again @0xc0170 @SenRamakri @jamesbeyond @offirko @kegilbert

offirko reviewed (commented):
Is there a  4_4_4 QPI read mode test?
Thanks

maciejbocianski commented:
@offirko
Currently `4_4_4` is not supported, but the test will evolve as new targets/flash chips support will be added, so there is no problem to add it in future when we will have stable version of test.

BTW adding new configs is very easy since everything is parameterized

maciejbocianski commented:
@offirko
It has been already done (see https://github.com/ARMmbed/mbed-os/pull/7325/commits/a3f76bcfa42a42b225ca4ab0cc7674cc76c07cdf) all per-target configuration like: `quad_enable`, `quad_disable`, `dual_enable`, `dual_disable`, `fast_mode_enable` was moved  to memory config files

jamesbeyond commented:
Can we have CI on this one, please ? @cmonr

cmonr commented:
/morph build

cmonr commented:
Trying that again.
/morph build

maciejbocianski committed:
hal-qspi test: fix QSPI preprocessor guard

maciejbocianski commented:
Test failed on some QSPI targets since only selected QSPI enabled targets support QSPI test.
Additional preprocessor guard has been added

maciejbocianski commented:
@cmonr @0xc0170 @adbridge 
morph can be re-triggered

0xc0170 commented:
/morph build

adbridge commented:
@studavekar Any idea why just one exporter failed with this :
```
download: s3://mbed-os/sources/7325/12c2d1a159e6b2840ecf0f05d7ac5f93af3108d8.tar.gz to .\12c2d1a159e6b2840ecf0f05d7ac5f93af3108d8.tar.gz
09:54:46 total 90352
09:54:46 drwxr-xr-x 1 Administrator 197121        0 Jul  6 09:54 .
09:54:46 drwxr-xr-x 1 Administrator 197121        0 Jul  6 09:54 ..
09:54:46 -rw-r--r-- 1 Administrator 197121 92513999 Jul  6 08:49 12c2d1a159e6b2840ecf0f05d7ac5f93af3108d8.tar.gz
09:54:46 drwxr-xr-x 1 Administrator 197121        0 Jul  6 09:54 DISCO_F469NI_iar
09:54:46 drwxr-xr-x 1 Administrator 197121        0 Jul  6 09:54 lockfile
09:54:51 [DISCO_F469NI_iar] $ env bash C:\Users\ADMINI~1\AppData\Local\Temp\jenkins1067965085058217093.sh
09:54:51 FATAL: command execution failed
09:54:51 Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to ec2-windows-exporter-iar (i-01ffcd9a0128d3cc6)
```
??

maciejbocianski commented:
lp_ticker fail on NRF51, looks not related (@mprse  will look in to it on monday)

@cmonr @0xc0170 @adbridge
can we restart morph

adbridge commented:
@maciejbocianski  We still need to work out why exporters failed too but for now I will kick both off again to see if they were just CI infrastructure issues
/morph export-build
/morph test

cmonr commented:
Interesting.
The test failure looks like it was a spurious test issue.

The export failure looks like ie was related to the CI node.

maciejbocianski commented:
Problem was already identified and @mprse will provide fix on Monday
```
[1530875539.22][CONN][RXD] >>> Running case #6: 'Test: LowPowerTimer - time measurement 1 ms.'...
[1530875539.22][CONN][RXD] :311::FAIL: Expected 0.001000 Was 0.001709
```

mprse commented:
> Problem was already identified and @mprse will provide fix on Monday

Fix can be found here: PR https://github.com/ARMmbed/mbed-os/pull/7471.

jamesbeyond commented:
Hi @adbridge @0xc0170, 

the bug cause the test failing got fixed by #7471, and it got merged now.
Can we re-trigger the test again on this PR ?

cmonr commented:
@jamesbeyond Before we restart, we'll need the feature branch to be rebased in order to pull in the changes from master.

`This branch is 59 commits ahead, 201 commits behind master.`

cmonr commented:
@maciejbocianski Was the close/reopen a mistake?

maciejbocianski commented:
It was mistake

maciejbocianski commented:
feature-qspi branch ant this PR  have been rebased

0xc0170 commented:
/morph build

adustm commented:
Hello @maciejbocianski 
Could you give me more details about the failure in case you remove the STM_WRITE_4IO_BUG_WORKAROUND ?
I don't see the difference with and without it.
Kind regards

cmonr commented:
/morph build

cmonr commented:
Not sure why this ended up timing out, but please take a look at the build errors.

maciejbocianski committed:
nrf52x: fix QSPI enable flag

QSPI feature was mistakenly moved form target nrf52840 to nrf52832
while rebasing. This change fixes it

maciejbocianski commented:
There was mistake while rebasing. `QSPI` flag was mistakenly moved from `nrf52840` to 'nrf52832` config.
Thus all nrf52832 based targets builds failed.
Fix was added: https://github.com/ARMmbed/mbed-os/pull/7325/commits/af90f2caa9daac83e2c55a7353217afe1c9a33df

cmonr commented:
@maciejbocianski Been there. It happens.

/morpg build

cmonr commented:
/morph build

maciejbocianski commented:
@adustm 
after rebasing feature_qspi branch to master, problems with malloc and with write delay seems to be not present anymore

adustm commented:
great news. The only remaining problem I can see is the ARM toolchain
```
---------------------+---------------------+-------------------------------------------------------+--------+--------+---------+--------------------+
 platform_name       | test suite          | test case                                             | passed | failed | result  | elapsed_time (sec) |
---------------------+---------------------+-------------------------------------------------------+--------+--------+---------+--------------------+
 DISCO_L475VG_IOT01A | tests-mbed_hal-qspi | qspi frequency setting test                           | 0      | 0      | SKIPPED | 0.0                |
 DISCO_L475VG_IOT01A | tests-mbed_hal-qspi | qspi init/free test                                   | 0      | 1      | FAIL    | 0.08               |
 DISCO_L475VG_IOT01A | tests-mbed_hal-qspi | qspi memory id test                                   | 1      | 0      | OK      | 0.1                |
 DISCO_L475VG_IOT01A | tests-mbed_hal-qspi | qspi write(1-1-1)/x1  read(1-1-1)/x1  repeat/x1  test | 0      | 0      | SKIPPED | 0.0                |
```

I can't find the fix for that at the moment (the issue disappears when opened with a debugger)

maciejbocianski commented:
On my side problem with ARM compiler also disappeared (after update to master)

adustm commented:
> On my side problem with ARM compiler also disappeared (after update to master)
@maciejbocianski did you rebase again or are you using feature-qspi branch that was rebased a few days ago ?

maciejbocianski commented:
@adustm there was only one rebase a few days ago, and I'm using it  (pure https://github.com/ARMmbed/mbed-os/tree/feature-qspi  plus yours PR #7581)

adustm commented:
Hello,
I'm using the same version as you. It's a matter of chance that yours is working.
I've found the bug.
qspi object is a local variable in qspi_init_free_test function (then located in the stack that contains whatever values). 
2 variables in qspi struct shall be initalized before the call to HAL_QSPI_Init function, inside qspi_init function. Otherwise you can be blocked at HAL_LOCK(hqspi) in stm32l4xx_hal_qspi.c, line 276
which was my case

```
 qspi_status_t qspi_init(qspi_t *obj, PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel, uint32_t hz, uint8_t mode)
 {
     // Enable interface clock for QSPI
      __HAL_RCC_QSPI_CLK_ENABLE();
 
     // Reset QSPI
     __HAL_RCC_QSPI_FORCE_RESET();
     __HAL_RCC_QSPI_RELEASE_RESET();
 
+    // Reset handle internal state
+    obj->handle.State = HAL_QSPI_STATE_RESET;
+    obj->handle.Lock = HAL_UNLOCKED;
+

```

PR#7586 has just arrived ;)

Kind regards
Armelle
